File: src/artifacts/strategy_operations.py
"""Build the governed current-state read model consumed by Strategy Console."""

from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import Any

from src.artifacts.model_run_bundle_v2 import validate_catalog
from src.artifacts.strategy_signal_ledger import (
    StrategySignalLedgerError,
    read_latest_evaluation,
)
from src.factors.strategy_snapshot import (
    StrategyFactorSnapshotError,
    validate_strategy_factor_snapshot,
)

SCHEMA_VERSION = "2.0.0"
QQQ_FAMILY = "qqq_rotation"
BYD_FAMILY = "byd_allocation"
US_RANKER_FAMILY = "us_ranker"
CN_RANKER_FAMILY = "cn_ranker"
RANKER_FAMILIES = {US_RANKER_FAMILY, CN_RANKER_FAMILY}
SUPPORTED_SIGNAL_FAMILIES = {QQQ_FAMILY, BYD_FAMILY, *RANKER_FAMILIES}
STATUS_VALUES = {
    "pipeline_unavailable",
    "awaiting_observation",
    "current_no_change",
    "target_pending_execution",
    "execution_observed",
    "stale",
    "blocked",
    "delivery_failed",
}
FRESHNESS_VALUES = {"current", "stale", "blocked", "unknown"}
QQQ_STATE_LABELS = {0: "Defensive", 1: "Transition", 2: "Risk-on"}
BYD_MODE_LABELS = {
    "defense": "Defensive",
    "offense": "Offense",
    "convex_expansion": "Convex expansion",
}


class StrategyOperationsError(ValueError):
    """Raised when a governed operations snapshot cannot be constructed."""


def _finite(value: object) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        result = float(value)
        if result == result and abs(result) != float("inf"):
            return result
    return None


def _weights(value: object) -> dict[str, float]:
    if not isinstance(value, Mapping):
        return {}
    result: dict[str, float] = {}
    for key, raw in value.items():
        number = _finite(raw)
        if number is None:
            raise StrategyOperationsError(
                f"allocation weight for {key!r} must be finite"
            )
        result[str(key)] = number
    return result


def _allocations(current: object, target: object) -> list[dict[str, object]]:
    current_weights = _weights(current)
    target_weights = _weights(target)
    assets = sorted(set(current_weights) | set(target_weights))
    return [
        {
            "asset": asset,
            "current": current_weights.get(asset, 0.0),
            "target": target_weights.get(asset, 0.0),
            "delta": target_weights.get(asset, 0.0) - current_weights.get(asset, 0.0),
        }
        for asset in assets
    ]


def _has_change(allocations: Sequence[Mapping[str, object]]) -> bool:
    return any(abs(float(row["delta"])) > 1e-9 for row in allocations)


def _formal_records(formal_catalog: Path) -> list[dict[str, Any]]:
    payload = json.loads(formal_catalog.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise StrategyOperationsError("formal catalog root must be an object")
    validate_catalog(payload)
    if payload.get("channel") != "formal":
        raise StrategyOperationsError("operations read model requires the formal catalog")
    if payload.get("research_only") is not True or payload.get("trade_ready") is not False:
        raise StrategyOperationsError("formal catalog research boundary is invalid")
    records = payload.get("records")
    if not isinstance(records, list) or not records:
        raise StrategyOperationsError("formal catalog contains no records")
    return [dict(record) for record in records if isinstance(record, Mapping)]


def _cadence(record: Mapping[str, Any]) -> tuple[str, str]:
    family = str(record.get("model_family_id", ""))
    if family == QQQ_FAMILY:
        return (
            "Every completed US market session",
            "Evaluate at close; target applies to the next eligible open.",
        )
    if family == BYD_FAMILY:
        return (
            "Every completed CN market session",
            "Evaluate after the common close; target applies to the next independently confirmed eligible open.",
        )
    if family in RANKER_FAMILIES:
        return (
            "Every 10 provider sessions",
            "Publish only on the governed rebalance session; target applies to the next eligible open.",
        )
    return (
        "Governed model cadence",
        "No governed current-target publisher is registered yet.",
    )


def _source(
    record: Mapping[str, Any], ledger: Mapping[str, Any] | None
) -> dict[str, object]:
    result: dict[str, object] = {
        "formal_bundle_id": record.get("bundle_id"),
        "formal_run_id": record.get("run_id"),
        "formal_evidence_cutoff": record.get("evidence_cutoff"),
        "ledger_fingerprint": None,
        "signal_sha256": None,
        "factor_catalog_implementation_hash": None,
        "workflow_run_id": None,
        "commit_sha": None,
        "github_issue_number": None,
    }
    if ledger is None:
        return result
    delivery = (
        ledger.get("delivery") if isinstance(ledger.get("delivery"), Mapping) else {}
    )
    workflow = (
        ledger.get("workflow") if isinstance(ledger.get("workflow"), Mapping) else {}
    )
    signal = ledger.get("signal") if isinstance(ledger.get("signal"), Mapping) else {}
    factor_evidence = (
        signal.get("factor_evidence")
        if isinstance(signal.get("factor_evidence"), Mapping)
        else {}
    )
    result.update(
        {
            "ledger_fingerprint": ledger.get("fingerprint"),
            "signal_sha256": ledger.get("signal_sha256"),
            "factor_catalog_implementation_hash": factor_evidence.get(
                "catalog_implementation_hash"
            ),
            "workflow_run_id": workflow.get("run_id"),
            "commit_sha": workflow.get("commit_sha"),
            "github_issue_number": delivery.get("github_issue_number"),
        }
    )
    return result


def _unavailable(record: Mapping[str, Any], *, awaiting: bool) -> dict[str, object]:
    cadence, next_policy = _cadence(record)
    if awaiting:
        status = "awaiting_observation"
        state_label = "Awaiting first governed evaluation"
        note = (
            "The signal publisher exists, but no cutoff-bound evaluation has been committed yet."
        )
    else:
        status = "pipeline_unavailable"
        state_label = "No governed current-target publisher"
        note = (
            "Formal historical evidence is available; live target state is intentionally unavailable."
        )
    return {
        "model_version_id": str(record["model_version_id"]),
        "status": status,
        "as_of": None,
        "latest_completed_session": record.get("evidence_cutoff"),
        "decision_cadence": cadence,
        "next_decision_policy": next_policy,
        "state_label": state_label,
        "decision_reason": note,
        "allocations": [],
        "turnover": None,
        "estimated_cost": None,
        "data_freshness": "unknown",
        "factor_freshness": "blocked" if awaiting else "unknown",
        "delivery_status": "not available",
        "source_label": (
            "Governed signal ledger" if awaiting else "Formal evidence only"
        ),
        "source_href": None,
        "note": note,
        "factor_evidence": [],
        "source_identity": _source(record, None),
    }


def _delivery(record: Mapping[str, Any]) -> tuple[str, int | None]:
    delivery = record.get("delivery")
    if not isinstance(delivery, Mapping):
        return "not available", None
    status = str(delivery.get("status") or "not available")
    issue = delivery.get("github_issue_number")
    return (
        status,
        int(issue) if isinstance(issue, int) and not isinstance(issue, bool) else None,
    )


def _factor_snapshot(
    signal: Mapping[str, Any], *, latest_data_date: object
) -> tuple[str, list[dict[str, Any]], str | None]:
    payload = signal.get("factor_evidence")
    if payload is None:
        return "blocked", [], "Latest signal predates the canonical factor snapshot contract."
    try:
        validate_strategy_factor_snapshot(payload)
    except StrategyFactorSnapshotError as exc:
        return "blocked", [], f"Canonical factor snapshot failed validation: {exc}"
    if not isinstance(payload, Mapping):
        return "blocked", [], "Canonical factor snapshot is not an object."
    if payload.get("observation_cutoff") != latest_data_date:
        return (
            "blocked",
            [],
            "Canonical factor snapshot cutoff does not match the signal data cutoff.",
        )
    rows = payload.get("factors")
    if not isinstance(rows, list):
        return "blocked", [], "Canonical factor snapshot has no factor rows."
    freshness = str(payload.get("freshness"))
    return freshness, [dict(row) for row in rows if isinstance(row, Mapping)], None


def _qqq_state(signal: Mapping[str, Any]) -> tuple[int, int, str]:
    current = signal.get("current_state", signal.get("current_formal_state", -1))
    target = signal.get("target_state", signal.get("target_formal_state", -1))
    current_state = (
        int(current)
        if isinstance(current, (int, float)) and not isinstance(current, bool)
        else -1
    )
    target_state = (
        int(target)
        if isinstance(target, (int, float)) and not isinstance(target, bool)
        else -1
    )
    from_label = QQQ_STATE_LABELS.get(current_state, f"State {current_state}")
    to_label = QQQ_STATE_LABELS.get(target_state, f"State {target_state}")
    current_overlay = signal.get("current_overlay")
    target_overlay = signal.get("target_overlay")
    if isinstance(target_overlay, str) and target_overlay:
        overlay = (
            target_overlay
            if current_overlay == target_overlay
            else f"{current_overlay or 'base'} → {target_overlay}"
        )
        label = f"{to_label} · {overlay}"
    else:
        label = (
            to_label if current_state == target_state else f"{from_label} → {to_label}"
        )
    return current_state, target_state, label


def _status(
    *,
    delivery_status: str,
    data_fresh: bool,
    factor_freshness: str,
    changed: bool,
) -> str:
    if delivery_status == "failed":
        return "delivery_failed"
    if not data_fresh or factor_freshness == "stale":
        return "stale"
    if factor_freshness != "current":
        return "blocked"
    return "target_pending_execution" if changed else "current_no_change"


def _qqq(record: Mapping[str, Any], ledger: Mapping[str, Any]) -> dict[str, object]:
    signal = ledger.get("signal")
    if not isinstance(signal, Mapping):
        raise StrategyOperationsError("QQQ ledger signal is missing")
    allocations = _allocations(signal.get("current_weights"), signal.get("target_weights"))
    changed = _has_change(allocations)
    data_fresh = signal.get("data_freshness_ok") is True
    delivery_status, issue_number = _delivery(ledger)
    latest = signal.get("latest_data_date") or ledger.get("latest_data_date")
    factor_freshness, factors, factor_error = _factor_snapshot(
        signal, latest_data_date=latest
    )
    _, _, state_label = _qqq_state(signal)
    cadence, next_policy = _cadence(record)
    decision_reason = signal.get("decision_reason_label") or signal.get(
        "decision_reason"
    )
    if not decision_reason:
        current_overlay = signal.get("current_overlay")
        target_overlay = signal.get("target_overlay")
        decision_reason = (
            f"{current_overlay or 'base'} → {target_overlay}"
            if isinstance(target_overlay, str) and current_overlay != target_overlay
            else str(target_overlay or "Frozen QQQ rotation evaluation.")
        )
    return {
        "model_version_id": str(record["model_version_id"]),
        "status": _status(
            delivery_status=delivery_status,
            data_fresh=data_fresh,
            factor_freshness=factor_freshness,
            changed=changed,
        ),
        "as_of": signal.get("signal_date"),
        "latest_completed_session": latest,
        "decision_cadence": cadence,
        "next_decision_policy": next_policy,
        "state_label": state_label,
        "decision_reason": str(decision_reason),
        "allocations": allocations,
        "turnover": _finite(signal.get("turnover_units")),
        "estimated_cost": _finite(signal.get("estimated_transaction_cost")),
        "data_freshness": "current" if data_fresh else "stale",
        "factor_freshness": factor_freshness,
        "delivery_status": delivery_status,
        "source_label": "Governed QQQ signal ledger",
        "source_href": (
            f"https://github.com/liuh886/alpha_engine/issues/{issue_number}"
            if issue_number
            else None
        ),
        "note": factor_error
        or (
            "Target is awaiting next-open execution evidence."
            if changed
            else "Latest governed evaluation retained the existing allocation."
        ),
        "factor_evidence": factors,
        "source_identity": _source(record, ledger),
    }


def _byd(record: Mapping[str, Any], ledger: Mapping[str, Any]) -> dict[str, object]:
    signal = ledger.get("signal")
    if not isinstance(signal, Mapping):
        raise StrategyOperationsError("BYD ledger signal is missing")
    allocations = _allocations(signal.get("current_weights"), signal.get("target_weights"))
    changed = _has_change(allocations)
    data_fresh = signal.get("data_freshness_ok") is True
    delivery_status, issue_number = _delivery(ledger)
    latest = signal.get("latest_data_date") or ledger.get("latest_data_date")
    factor_freshness, factors, factor_error = _factor_snapshot(
        signal, latest_data_date=latest
    )
    mode = str(signal.get("target_mode") or "")
    cadence, next_policy = _cadence(record)
    return {
        "model_version_id": str(record["model_version_id"]),
        "status": _status(
            delivery_status=delivery_status,
            data_fresh=data_fresh,
            factor_freshness=factor_freshness,
            changed=changed,
        ),
        "as_of": signal.get("signal_date"),
        "latest_completed_session": latest,
        "decision_cadence": cadence,
        "next_decision_policy": next_policy,
        "state_label": BYD_MODE_LABELS.get(mode, mode or "BYD allocation"),
        "decision_reason": signal.get("transition_label")
        or signal.get("transition_type")
        or "Frozen BYD allocation evaluation.",
        "allocations": allocations,
        "turnover": _finite(signal.get("turnover_units")),
        "estimated_cost": _finite(signal.get("estimated_transaction_cost")),
        "data_freshness": "current" if data_fresh else "stale",
        "factor_freshness": factor_freshness,
        "delivery_status": delivery_status,
        "source_label": "Governed BYD signal ledger",
        "source_href": (
            f"https://github.com/liuh886/alpha_engine/issues/{issue_number}"
            if issue_number
            else None
        ),
        "note": factor_error
        or (
            "Target allocation is published; brokerage execution is outside Alpha Engine."
            if changed
            else "Latest governed evaluation retained the existing allocation."
        ),
        "factor_evidence": factors,
        "source_identity": _source(record, ledger),
    }


def _ranker(record: Mapping[str, Any], ledger: Mapping[str, Any]) -> dict[str, object]:
    signal = ledger.get("signal")
    if not isinstance(signal, Mapping):
        raise StrategyOperationsError("ranker ledger signal is missing")
    allocations = _allocations(signal.get("current_weights"), signal.get("target_weights"))
    changed = _has_change(allocations)
    data_fresh = signal.get("data_freshness_ok") is True
    delivery_status, issue_number = _delivery(ledger)
    latest = signal.get("latest_data_date") or ledger.get("latest_data_date")
    factor_freshness, factors, factor_error = _factor_snapshot(
        signal, latest_data_date=latest
    )
    family = str(record.get("model_family_id", ""))
    diagnostics = signal.get("diagnostics") if isinstance(signal.get("diagnostics"), Mapping) else {}
    if family == US_RANKER_FAMILY:
        state_label = "US Top-15 rebalance"
    elif diagnostics.get("risk_on") is False:
        state_label = "CN risk-off · CSI300 fallback"
    else:
        state_label = "CN risk-on · sector 4×1"
    cadence, next_policy = _cadence(record)
    return {
        "model_version_id": str(record["model_version_id"]),
        "status": _status(
            delivery_status=delivery_status,
            data_fresh=data_fresh,
            factor_freshness=factor_freshness,
            changed=changed,
        ),
        "as_of": signal.get("signal_date"),
        "latest_completed_session": latest,
        "decision_cadence": cadence,
        "next_decision_policy": next_policy,
        "state_label": state_label,
        "decision_reason": str(signal.get("reason_code") or "Frozen 10-session ranker evaluation."),
        "allocations": allocations,
        "turnover": _finite(signal.get("turnover_units")),
        "estimated_cost": _finite(signal.get("estimated_transaction_cost")),
        "data_freshness": "current" if data_fresh else "stale",
        "factor_freshness": factor_freshness,
        "delivery_status": delivery_status,
        "source_label": "Governed 10-session ranker signal ledger",
        "source_href": (
            f"https://github.com/liuh886/alpha_engine/issues/{issue_number}"
            if issue_number
            else None
        ),
        "note": factor_error
        or (
            "Target is published for the next eligible open."
            if changed
            else "The governed rebalance retained the existing target."
        ),
        "factor_evidence": factors,
        "source_identity": _source(record, ledger),
    }


def build_operations_payload(
    *, formal_catalog: Path, ledger_root: Path, generated_at: str
) -> dict[str, object]:
    records: list[dict[str, object]] = []
    for formal in _formal_records(formal_catalog):
        model_version_id = str(formal["model_version_id"])
        family = str(formal.get("model_family_id", ""))
        try:
            ledger = read_latest_evaluation(
                ledger_root / model_version_id,
                model_version_id=model_version_id,
            )
        except (OSError, json.JSONDecodeError, StrategySignalLedgerError) as exc:
            blocked = _unavailable(
                formal, awaiting=family in SUPPORTED_SIGNAL_FAMILIES
            )
            blocked["status"] = "blocked"
            blocked["decision_reason"] = str(exc)
            blocked["note"] = (
                "Governed signal ledger failed validation; operations fail closed."
            )
            records.append(blocked)
            continue
        if ledger is None:
            records.append(
                _unavailable(formal, awaiting=family in SUPPORTED_SIGNAL_FAMILIES)
            )
        elif family == QQQ_FAMILY:
            records.append(_qqq(formal, ledger))
        elif family == BYD_FAMILY:
            records.append(_byd(formal, ledger))
        elif family in RANKER_FAMILIES:
            records.append(_ranker(formal, ledger))
        else:
            blocked = _unavailable(formal, awaiting=False)
            blocked["status"] = "blocked"
            blocked["decision_reason"] = (
                "A signal ledger exists but no governed operations adapter is registered."
            )
            blocked["note"] = (
                "Backend adapter required before this signal can be published."
            )
            records.append(blocked)

    return {
        "schema_version": SCHEMA_VERSION,
        "generated_at": generated_at,
        "research_only": True,
        "trade_ready": False,
        "records": records,
    }


def validate_operations_payload(payload: object) -> None:
    if not isinstance(payload, Mapping):
        raise StrategyOperationsError("operations payload root must be an object")
    if payload.get("schema_version") != SCHEMA_VERSION:
        raise StrategyOperationsError("unsupported operations schema")
    if payload.get("research_only") is not True or payload.get("trade_ready") is not False:
        raise StrategyOperationsError("operations research boundary is invalid")
    records = payload.get("records")
    if not isinstance(records, Sequence) or isinstance(records, (str, bytes)):
        raise StrategyOperationsError("operations records must be a list")
    seen: set[str] = set()
    for index, value in enumerate(records):
        if not isinstance(value, Mapping):
            raise StrategyOperationsError(
                f"operations record {index} must be an object"
            )
        model_id = value.get("model_version_id")
        if not isinstance(model_id, str) or not model_id:
            raise StrategyOperationsError(
                f"operations record {index} has no model identity"
            )
        if model_id in seen:
            raise StrategyOperationsError(f"duplicate operations model: {model_id}")
        seen.add(model_id)
        if value.get("status") not in STATUS_VALUES:
            raise StrategyOperationsError(f"invalid operations status for {model_id}")
        if (
            value.get("data_freshness") not in FRESHNESS_VALUES
            or value.get("factor_freshness") not in FRESHNESS_VALUES
        ):
            raise StrategyOperationsError(
                f"invalid freshness state for {model_id}"
            )
        factor_evidence = value.get("factor_evidence")
        if not isinstance(factor_evidence, Sequence) or isinstance(
            factor_evidence, (str, bytes)
        ):
            raise StrategyOperationsError(
                f"factor evidence must be a list for {model_id}"
            )
        if value.get("factor_freshness") == "current" and not factor_evidence:
            raise StrategyOperationsError(
                f"current factor freshness requires evidence for {model_id}"
            )

File: qlib-dashboard/src/lib/strategy-operations.ts
import type { GovernedRunSummary } from './governed-run';
import type { AccessTier } from './model-access';
import { assetUrl } from './runtime-capabilities';

export type StrategyOperationalStatus =
  | 'pipeline_unavailable'
  | 'awaiting_observation'
  | 'current_no_change'
  | 'target_pending_execution'
  | 'execution_observed'
  | 'stale'
  | 'blocked'
  | 'delivery_failed';

export type StrategyFreshness = 'current' | 'stale' | 'blocked' | 'unknown';
export type FactorEffect = 'support' | 'veto' | 'neutral';

export interface StrategyAllocationLeg {
  asset: string;
  current: number;
  target: number;
  delta: number;
}

export interface StrategyFactorEvidence {
  factorId: string;
  factorVersion: string;
  implementationHash: string;
  displayName: string;
  informationFamily: string;
  value: number | string | boolean;
  reference: unknown;
  state: string;
  effect: FactorEffect;
  reasonCode: string;
  observedAt: string;
}

export interface StrategySourceIdentity {
  formalBundleId: string | null;
  formalRunId: string | null;
  formalEvidenceCutoff: string | null;
  ledgerFingerprint: string | null;
  signalSha256: string | null;
  factorCatalogImplementationHash: string | null;
  workflowRunId: string | null;
  commitSha: string | null;
  githubIssueNumber: number | null;
}

export interface StrategyOperationsSnapshot {
  strategyId: string;
  modelVersionId: string;
  currentOperationsAccess: AccessTier;
  status: StrategyOperationalStatus;
  asOf: string | null;
  latestCompletedSession: string | null;
  decisionCadence: string;
  nextDecision: string;
  stateLabel: string;
  decisionReason: string;
  allocations: StrategyAllocationLeg[];
  turnover: number | null;
  estimatedCost: number | null;
  dataFreshness: StrategyFreshness;
  factorFreshness: StrategyFreshness;
  deliveryStatus: string;
  sourceLabel: string;
  sourceHref: string | null;
  note: string;
  factorEvidence: StrategyFactorEvidence[];
  sourceIdentity: StrategySourceIdentity;
}

interface OperationsDocument {
  schema_version?: unknown;
  research_only?: unknown;
  trade_ready?: unknown;
  records?: unknown;
}

interface ProtectedOperationRow {
  strategy_id?: unknown;
  model_version_id?: unknown;
  snapshot?: unknown;
}

interface ProtectedOperationResult {
  data: ProtectedOperationRow | null;
  error: { message?: string } | null;
}

interface ProtectedOperationFilter {
  eq: (column: string, value: string) => ProtectedOperationFilter;
  maybeSingle: () => Promise<ProtectedOperationResult>;
}

interface ProtectedOperationQuery {
  select: (columns: string) => ProtectedOperationFilter;
}

export interface StrategyOperationsClient {
  from: (table: string) => ProtectedOperationQuery;
}

const STATUS = new Set<StrategyOperationalStatus>([
  'pipeline_unavailable',
  'awaiting_observation',
  'current_no_change',
  'target_pending_execution',
  'execution_observed',
  'stale',
  'blocked',
  'delivery_failed',
]);
const FRESHNESS = new Set<StrategyFreshness>(['current', 'stale', 'blocked', 'unknown']);
const EFFECT = new Set<FactorEffect>(['support', 'veto', 'neutral']);
const ACCESS = new Set<AccessTier>(['public', 'authenticated', 'pro', 'owner']);

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) throw new Error(message);
}

function nullableString(value: unknown): string | null {
  return typeof value === 'string' && value.length > 0 ? value : null;
}

function nullableNumber(value: unknown): number | null {
  return typeof value === 'number' && Number.isFinite(value) ? value : null;
}

function parseFactorEvidence(value: unknown, id: string, index: number): StrategyFactorEvidence {
  assert(Boolean(value) && typeof value === 'object' && !Array.isArray(value), `Invalid factor evidence ${index} for ${id}`);
  const row = value as Record<string, unknown>;
  assert(typeof row.factor_id === 'string' && row.factor_id.length > 0, `Missing factor id for ${id}`);
  assert(typeof row.factor_version === 'string' && row.factor_version.length > 0, `Missing factor version for ${id}/${row.factor_id}`);
  assert(typeof row.implementation_hash === 'string' && /^[a-f0-9]{64}$/.test(row.implementation_hash), `Invalid factor implementation hash for ${id}/${row.factor_id}`);
  assert(typeof row.display_name === 'string' && row.display_name.length > 0, `Missing factor display name for ${id}/${row.factor_id}`);
  assert(typeof row.information_family === 'string' && row.information_family.length > 0, `Missing factor family for ${id}/${row.factor_id}`);
  assert(typeof row.state === 'string' && row.state.length > 0, `Missing factor state for ${id}/${row.factor_id}`);
  assert(EFFECT.has(row.effect as FactorEffect), `Invalid factor effect for ${id}/${row.factor_id}`);
  assert(typeof row.reason_code === 'string' && row.reason_code.length > 0, `Missing factor reason code for ${id}/${row.factor_id}`);
  assert(typeof row.observed_at === 'string' && row.observed_at.length > 0, `Missing factor observation date for ${id}/${row.factor_id}`);
  const observedValue = row.value;
  assert(
    (typeof observedValue === 'number' && Number.isFinite(observedValue)) || typeof observedValue === 'string' || typeof observedValue === 'boolean',
    `Invalid factor value for ${id}/${row.factor_id}`,
  );
  return {
    factorId: row.factor_id,
    factorVersion: row.factor_version,
    implementationHash: row.implementation_hash,
    displayName: row.display_name,
    informationFamily: row.information_family,
    value: observedValue,
    reference: row.reference,
    state: row.state,
    effect: row.effect as FactorEffect,
    reasonCode: row.reason_code,
    observedAt: row.observed_at,
  };
}

export function parseStrategyOperationsSnapshot(value: unknown): StrategyOperationsSnapshot {
  assert(Boolean(value) && typeof value === 'object' && !Array.isArray(value), 'Invalid strategy operations record');
  const record = value as Record<string, unknown>;
  const modelVersionId = record.model_version_id;
  const strategyId = record.strategy_id;
  assert(typeof modelVersionId === 'string' && modelVersionId.length > 0, 'Strategy operations model identity is missing');
  assert(typeof strategyId === 'string' && strategyId.length > 0, `Strategy identity is missing for ${modelVersionId}`);
  assert(ACCESS.has(record.current_operations_access as AccessTier), `Unsupported operations access for ${modelVersionId}`);
  assert(STATUS.has(record.status as StrategyOperationalStatus), `Unsupported operations status for ${modelVersionId}`);
  assert(FRESHNESS.has(record.data_freshness as StrategyFreshness), `Unsupported data freshness for ${modelVersionId}`);
  assert(FRESHNESS.has(record.factor_freshness as StrategyFreshness), `Unsupported factor freshness for ${modelVersionId}`);
  assert(typeof record.decision_cadence === 'string' && record.decision_cadence.length > 0, `Missing decision cadence for ${modelVersionId}`);
  assert(typeof record.next_decision_policy === 'string' && record.next_decision_policy.length > 0, `Missing next decision policy for ${modelVersionId}`);
  assert(typeof record.state_label === 'string' && record.state_label.length > 0, `Missing state label for ${modelVersionId}`);
  assert(typeof record.decision_reason === 'string' && record.decision_reason.length > 0, `Missing decision reason for ${modelVersionId}`);
  assert(typeof record.delivery_status === 'string' && record.delivery_status.length > 0, `Missing delivery status for ${modelVersionId}`);
  assert(typeof record.source_label === 'string' && record.source_label.length > 0, `Missing source label for ${modelVersionId}`);
  assert(typeof record.note === 'string' && record.note.length > 0, `Missing operations note for ${modelVersionId}`);
  assert(Array.isArray(record.allocations), `Missing allocations for ${modelVersionId}`);
  assert(Array.isArray(record.factor_evidence), `Missing factor evidence for ${modelVersionId}`);
  assert(Boolean(record.source_identity) && typeof record.source_identity === 'object' && !Array.isArray(record.source_identity), `Missing source identity for ${modelVersionId}`);

  const allocations = record.allocations.map((allocation, index) => {
    assert(Boolean(allocation) && typeof allocation === 'object' && !Array.isArray(allocation), `Invalid allocation ${index} for ${modelVersionId}`);
    const row = allocation as Record<string, unknown>;
    assert(typeof row.asset === 'string' && row.asset.length > 0, `Missing allocation asset for ${modelVersionId}`);
    assert(typeof row.current === 'number' && Number.isFinite(row.current), `Invalid current weight for ${modelVersionId}/${row.asset}`);
    assert(typeof row.target === 'number' && Number.isFinite(row.target), `Invalid target weight for ${modelVersionId}/${row.asset}`);
    assert(typeof row.delta === 'number' && Number.isFinite(row.delta), `Invalid allocation delta for ${modelVersionId}/${row.asset}`);
    return { asset: row.asset, current: row.current, target: row.target, delta: row.delta };
  });

  const factorEvidence = record.factor_evidence.map((row, index) => parseFactorEvidence(row, modelVersionId, index));
  if (record.factor_freshness === 'current') {
    assert(factorEvidence.length > 0, `Current factor freshness requires evidence for ${modelVersionId}`);
  }

  const source = record.source_identity as Record<string, unknown>;
  return {
    strategyId,
    modelVersionId,
    currentOperationsAccess: record.current_operations_access as AccessTier,
    status: record.status as StrategyOperationalStatus,
    asOf: nullableString(record.as_of),
    latestCompletedSession: nullableString(record.latest_completed_session),
    decisionCadence: record.decision_cadence,
    nextDecision: record.next_decision_policy,
    stateLabel: record.state_label,
    decisionReason: record.decision_reason,
    allocations,
    turnover: nullableNumber(record.turnover),
    estimatedCost: nullableNumber(record.estimated_cost),
    dataFreshness: record.data_freshness as StrategyFreshness,
    factorFreshness: record.factor_freshness as StrategyFreshness,
    deliveryStatus: record.delivery_status,
    sourceLabel: record.source_label,
    sourceHref: nullableString(record.source_href),
    note: record.note,
    factorEvidence,
    sourceIdentity: {
      formalBundleId: nullableString(source.formal_bundle_id),
      formalRunId: nullableString(source.formal_run_id),
      formalEvidenceCutoff: nullableString(source.formal_evidence_cutoff),
      ledgerFingerprint: nullableString(source.ledger_fingerprint),
      signalSha256: nullableString(source.signal_sha256),
      factorCatalogImplementationHash: nullableString(source.factor_catalog_implementation_hash),
      workflowRunId: nullableString(source.workflow_run_id),
      commitSha: nullableString(source.commit_sha),
      githubIssueNumber: typeof source.github_issue_number === 'number' && Number.isInteger(source.github_issue_number)
        ? source.github_issue_number
        : null,
    },
  };
}

function blocked(run: GovernedRunSummary, message: string): StrategyOperationsSnapshot {
  return {
    strategyId: run.modelVersionId,
    modelVersionId: run.modelVersionId,
    currentOperationsAccess: 'public',
    status: 'blocked',
    asOf: null,
    latestCompletedSession: run.evidenceCutoff || null,
    decisionCadence: 'Operations read model unavailable',
    nextDecision: 'Restore the governed operations publication before using current-state claims.',
    stateLabel: 'Operating data blocked',
    decisionReason: message,
    allocations: [],
    turnover: null,
    estimatedCost: null,
    dataFreshness: 'blocked',
    factorFreshness: 'blocked',
    deliveryStatus: 'not available',
    sourceLabel: 'Governed operations snapshot',
    sourceHref: null,
    note: message,
    factorEvidence: [],
    sourceIdentity: {
      formalBundleId: null,
      formalRunId: run.runId,
      formalEvidenceCutoff: run.evidenceCutoff || null,
      ledgerFingerprint: null,
      signalSha256: null,
      factorCatalogImplementationHash: null,
      workflowRunId: null,
      commitSha: null,
      githubIssueNumber: null,
    },
  };
}

async function fetchOperations(): Promise<Map<string, StrategyOperationsSnapshot>> {
  const response = await fetch(assetUrl('data/strategy-operations/snapshots.json'), { cache: 'no-store' });
  if (!response.ok) throw new Error(`Strategy operations snapshot unavailable (${response.status})`);
  const value = await response.json() as OperationsDocument;
  assert(value.schema_version === '2.1.0', 'Unsupported strategy operations schema');
  assert(value.research_only === true && value.trade_ready === false, 'Invalid strategy operations boundary');
  assert(Array.isArray(value.records), 'Strategy operations records are missing');
  const snapshots = value.records.map(parseStrategyOperationsSnapshot);
  assert(new Set(snapshots.map((snapshot) => snapshot.modelVersionId)).size === snapshots.length, 'Duplicate strategy operations model identity');
  assert(new Set(snapshots.map((snapshot) => snapshot.strategyId)).size === snapshots.length, 'Duplicate stable strategy identity');
  return new Map(snapshots.map((snapshot) => [snapshot.modelVersionId, snapshot]));
}

export async function loadProtectedStrategyOperation(
  client: StrategyOperationsClient,
  strategyId: string,
  modelVersionId: string,
): Promise<StrategyOperationsSnapshot | null> {
  const { data, error } = await client
    .from('strategy_operation_snapshots')
    .select('strategy_id,model_version_id,snapshot')
    .eq('strategy_id', strategyId)
    .eq('model_version_id', modelVersionId)
    .maybeSingle();
  if (error) throw new Error(error.message || 'Protected strategy operations request failed.');
  if (!data) return null;
  assert(data.strategy_id === strategyId, 'Protected strategy identity mismatch');
  assert(data.model_version_id === modelVersionId, 'Protected model identity mismatch');
  const snapshot = parseStrategyOperationsSnapshot(data.snapshot);
  assert(snapshot.strategyId === strategyId, 'Protected snapshot strategy identity mismatch');
  assert(snapshot.modelVersionId === modelVersionId, 'Protected snapshot model identity mismatch');
  return snapshot;
}

export async function loadStrategyOperations(runs: GovernedRunSummary[]): Promise<Map<string, StrategyOperationsSnapshot>> {
  const formalRuns = runs.filter((run) => run.channel === 'formal');
  try {
    const snapshots = await fetchOperations();
    const missingIds = formalRuns
      .map((run) => run.modelVersionId)
      .filter((modelVersionId) => !snapshots.has(modelVersionId));
    assert(missingIds.length === 0, `Strategy operations are missing accepted formal models: ${missingIds.join(', ')}`);
    return new Map(formalRuns.map((run) => [run.modelVersionId, snapshots.get(run.modelVersionId)!]));
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Strategy operations snapshot is unavailable.';
    return new Map(formalRuns.map((run) => [run.modelVersionId, blocked(run, message)]));
  }
}

export const STRATEGY_STATUS_LABEL: Record<StrategyOperationalStatus, string> = {
  pipeline_unavailable: 'Signal unavailable',
  awaiting_observation: 'Awaiting observation',
  current_no_change: 'Current · no change',
  target_pending_execution: 'New target',
  execution_observed: 'Execution observed',
  stale: 'Stale data',
  blocked: 'Operating data blocked',
  delivery_failed: 'Delivery failed',
};
